### ISSUE TITLE
[FR] Support fallback for `find_files` in archives

### ISSUE THREAD

**Issue by abravalheri** [pypa/setuptools-scm#751]
Hello. I noticed very recently that the latest versions of `setuptools_scm` support git archives via the very clever `.git_archival.txt` trick. That is awesome 👏, thank you very much for that!

I also noticed that this trick will only work as a fallback for the lack of tags to derive versions...
There is a second aspect why people use `setuptools_scm` (the `file_finders` hook), and unfortunately that does not seem to be covered by `.git_archival.txt`.

This impacts people's ability to omit `MANIFEST.in` and/or use `include_package_data=True` with setuptools if they want to support downstream re-packagers that rely on archives instead of sdists.

I understand that this is a known limitation, but I think we can do something around it:

### Proposal

Let's simply list all existing files under the project folder when `setuptools_scm.integration:find_files` is called from a git/hg archive as a fallback.

The rationale is: if someone is building from an archive, all the files that should have been ignored by git/hg were already removed when the archive was being created...

To ensure that this fallback only runs if an archive is given we can do the following:

1. Don't run any fallback file finder if any of the original hooks can run (i.e. the project contain .git/.hg folders and git/hg are able to list files).
2. Check for the presence of `.git_archival.txt` files. More specifically check if one of the `$Format:...$` strings were expanded, e.g. for `node`.
3. Check for the presence of `.hg_archival.txt` (we can add something else here, e.g. the file must not be empty?)

I will create a PR with the implementation.

### Reproducer for the current circumstances

Let's create a project that uses `include_package_data=True` and does not contain a `MANIFEST.in`.
We then try to build it from the git archive:

```bash
rm -rf /tmp/myproj /tmp/myproj-from-archive /tmp/myproj.zip
mkdir -p /tmp/myproj/src/myproj
cd /tmp/myproj

touch src/myproj/__init__.py
touch src/myproj/file.json

cat <<EOF > pyproject.toml
[build-system]
requires = ["setuptools", "setuptools-scm[toml]"]
build-backend = "setuptools.build_meta"

[tool.setuptools_scm]
version_scheme = "no-guess-dev"
EOF

cat <<EOF > setup.cfg
[metadata]
name = myproj

[options]
include_package_data = True
packages = find:
package_dir =
    =src

[options.packages.find]
where = src
EOF

cat <<EOF > .git_archival.txt
node: \$Format:%H\$
node-date: \$Format:%cI\$
describe-name: \$Format:%(describe:tags=true)\$
ref-names: \$Format:%D\$
EOF

cat <<EOF > .gitattributes
.git_archival.txt  export-subst
EOF

git init .
git add .
git commit -m "Initial Commit"


git archive --format zip HEAD -o /tmp/myproj.zip
unzip /tmp/myproj.zip -d /tmp/myproj-from-archive
cd /tmp/myproj-from-archive

tree
# .
# ├── pyproject.toml
# ├── setup.cfg
# └── src
#     └── myproj
#         ├── __init__.py
#         └── file.json

pipx run build
unzip -l dist/*.whl

# Archive:  dist/myproj-0.0-py3-none-any.whl
#   Length      Date    Time    Name
# ---------  ---------- -----   ----
#         0  2022-07-15 11:30   myproj/__init__.py
#        49  2022-07-15 11:31   myproj-0.0.dist-info/METADATA
#        92  2022-07-15 11:31   myproj-0.0.dist-info/WHEEL
#         7  2022-07-15 11:31   myproj-0.0.dist-info/top_level.txt
#       355  2022-07-15 11:31   myproj-0.0.dist-info/RECORD
# ---------                     -------
#       503                     5 files
```

Here we can see that the output wheel *does not* contain the `.json` file.

At the same time, if we build the original repository, we can see that it *does* contain the `.json` file (as intended):

```bash
cd /tmp/myproj
pipx run build
unzip -l dist/*.whl
# Archive:  dist/myproj-0.0.post1.dev1+geee6468-py3-none-any.whl
#   Length      Date    Time    Name
# ---------  ---------- -----   ----
#         0  2022-07-15 11:30   myproj/__init__.py
#         0  2022-07-15 11:30   myproj/file.json
#        69  2022-07-15 11:32   myproj-0.0.post1.dev1+geee6468.dist-info/METADATA
#        92  2022-07-15 11:32   myproj-0.0.post1.dev1+geee6468.dist-info/WHEEL
#         7  2022-07-15 11:32   myproj-0.0.post1.dev1+geee6468.dist-info/top_level.txt
#       505  2022-07-15 11:32   myproj-0.0.post1.dev1+geee6468.dist-info/RECORD
# ---------                     -------
#       673                     6 files
```

**Comment by RonnyPfannschmidt** [comment]
:+1:

**Comment by RonnyPfannschmidt** [comment]
closing as solved by #752
